# 딥러닝 동작원리. update rule
1. 경사하강법. gradient decent
2. 경사하강법 + 변수추가 --- 다중선형회귀
3. 로지스틱 회귀. logistic regression

by tensorflow

In [1]:
import tensorflow as tf

C:\Users\kitcoop\Anaconda3\lib\site-packages\h5py\__init__.py:34: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters


## 1

### 경사하강법
미분한 기울기값을 비교하며 loss오차 최솟값 찾기

In [5]:
data= [[2,81], [4,93], [6,91], [8,97]]
x_data= [x_row[0] for x_row in data] #요소별 앞값: 공부시간
y_data= [y_row[1] for y_row in data] #요소별 뒷값: 시험성적 실제값

임의 최솟값. 기울기 a, 절편 b
- a [0,10]
- b [0,100]

In [6]:
a= tf.Variable(tf.random_uniform([1], 0, 10, dtype= tf.float64, seed= 0))
b= tf.Variable(tf.random_uniform([1], 0, 100,dtype= tf.float64, seed= 0))

y= ax+ b

In [7]:
y= a* x_data+ b

tf.RMSE

In [8]:
rmse= tf.sqrt(tf.reduce_mean(tf.square(y- y_data)))

학습률

In [9]:
learning_rate= 0.1

RMSE 최솟값

In [11]:
gradient_decent= tf.train.\
    GradientDescentOptimizer(learning_rate).minimize(rmse)
# GradientDescentOptimizer: 경사하강법

##### tf로 학습
갈수록 RMSE가 줄어듦. a,b는 수렴

In [18]:
with tf.Session() as sess: # 세션: 학습위한 자원할당
    sess.run(tf.global_variables_initializer()) #변수 초기화
    for step in range(2001): #2001-0+1 번 실행
        sess.run(gradient_decent)
        if 0== step%100: #100번째 마다 보자
            print("epoch%12.6f, RMSE%10.6f, 기울기a%10.6f, y절편b%10.6f"%
                 (step, sess.run(rmse), sess.run(a), sess.run(b)))

epoch    0.000000, RMSE 30.213942, 기울기a  7.523466, y절편b 80.598359
epoch  100.000000, RMSE  2.886024, 기울기a  2.229946, y절편b 79.418055
epoch  200.000000, RMSE  2.882609, 기울기a  2.260135, y절편b 79.237895
epoch  300.000000, RMSE  2.881502, 기울기a  2.277324, y절편b 79.135320
epoch  400.000000, RMSE  2.881143, 기울기a  2.287103, y절편b 79.076963
epoch  500.000000, RMSE  2.881027, 기울기a  2.292665, y절편b 79.043771
epoch  600.000000, RMSE  2.880990, 기울기a  2.295829, y절편b 79.024893
epoch  700.000000, RMSE  2.880978, 기울기a  2.297628, y절편b 79.014157
epoch  800.000000, RMSE  2.880974, 기울기a  2.298651, y절편b 79.008051
epoch  900.000000, RMSE  2.880973, 기울기a  2.299233, y절편b 79.004579
epoch 1000.000000, RMSE  2.880972, 기울기a  2.299564, y절편b 79.002604
epoch 1100.000000, RMSE  2.880972, 기울기a  2.299752, y절편b 79.001481
epoch 1200.000000, RMSE  2.880972, 기울기a  2.299859, y절편b 79.000842
epoch 1300.000000, RMSE  2.880972, 기울기a  2.299920, y절편b 79.000479
epoch 1400.000000, RMSE  2.880972, 기울기a  2.299954, y절편b 79.000272
epoch 1500

## 2

### 경사하강법 + 변수추가
6시간 공부하면 93점일줄 알았는데 91점. 예측값 != 실제값
> 공부시간 외 다른 요소가 있군! 새 예측하려면 변수 개수를 늘려야 -> 다중회귀

In [3]:
data=[[2,0,81], [4,4,93], [6,2,91], [8,3,97]]
x1= [x_row1[0] for x_row1 in data]
x2= [x_row2[1] for x_row2 in data] # 변수 추가
y_data= [y_row[2] for y_row in data] #data에서 elem하나씩 꺼내와 y_row에 넣는데, 그중 3번째 것들만

임의 초깃값 기울기a [0,10], 절편b [0,100]

In [4]:
# a2: 추가 변수
a1= tf.Variable(tf.random_uniform([1], 0,10, dtype= tf.float64, seed= 0))
a2= tf.Variable(tf.random_uniform([1], 0,10, dtype= tf.float64, seed= 0))
b= tf.Variable(tf.random_uniform([1], 0,100, dtype= tf.float64, seed= 0))

In [5]:
y= a1*x1 + a2*x2 + b

y: 예측값, y_data: 실제값

In [9]:
rmse= tf.sqrt(tf.reduce_mean(tf.square( y- y_data)))

In [10]:
learning_rate= 0.1

###### 경사하강

In [11]:
gradient_decent= tf.train.GradientDescentOptimizer(learning_rate).minimize(rmse)

###### 학습
100회째부터 확 수렴하는걸 보니 초깃값에 그리 큰 영향받진 X

- 공부시간1개: RMSE 2.880972
- 시간+다른변수: RMSE 1.837006

In [14]:
with tf.Session() as ss:
    ss.run(tf.global_variables_initializer())
    for step in range(2001):
        ss.run(gradient_decent)
        if 0== step%100:
            print('epoch%12.6f, RMSE%10.6f, 기울기a1%10.6f, 기울기a2%10.6f, y절편b%10.6f'%
                 (step, ss.run(rmse), ss.run(a1), ss.run(a2), ss.run(b)))

epoch    0.000000, RMSE 49.184210, 기울기a1  7.527029, 기울기a2  7.816014, y절편b 80.597994
epoch  100.000000, RMSE  1.836812, 기울기a1  1.130559, 기울기a2  2.131645, y절편b 78.511887
epoch  200.000000, RMSE  1.836972, 기울기a1  1.187870, 기울기a2  2.148749, y절편b 78.105703
epoch  300.000000, RMSE  1.837000, 기울기a1  1.212219, 기울기a2  2.157104, y절편b 77.935173
epoch  400.000000, RMSE  1.837005, 기울기a1  1.222564, 기울기a2  2.160666, y절편b 77.863553
epoch  500.000000, RMSE  1.837006, 기울기a1  1.226930, 기울기a2  2.162172, y절편b 77.833469
epoch  600.000000, RMSE  1.837006, 기울기a1  1.228768, 기울기a2  2.162807, y절편b 77.820831
epoch  700.000000, RMSE  1.837006, 기울기a1  1.229541, 기울기a2  2.163073, y절편b 77.815522
epoch  800.000000, RMSE  1.837006, 기울기a1  1.229866, 기울기a2  2.163186, y절편b 77.813291
epoch  900.000000, RMSE  1.837006, 기울기a1  1.230002, 기울기a2  2.163233, y절편b 77.812354
epoch 1000.000000, RMSE  1.837006, 기울기a1  1.230059, 기울기a2  2.163253, y절편b 77.811961
epoch 1100.000000, RMSE  1.837006, 기울기a1  1.230083, 기울기a2  2.163261, y절편b 77

## 3

### 로지스틱 회귀
- return T or F 
- param (ax + b)
    - a 경사도. oo이면 각짐
    - b 좌우로 이동. 크면 우측으로

#### 시그모이드: 그냥한번 직접 짜봄

In [16]:
import numpy as np

In [17]:
data= [[2,0], [4,0], [6,0], [8,1], [10,1], [12,1], [14,1]]
x_data= [x_row[0] for x_row in data]
y_data= [y_row[1] for y_row in data]

In [18]:
#임의 초깃값
a= tf.Variable(tf.random_normal([1], dtype= tf.float64, seed= 0))
b= tf.Variable(tf.random_normal([1], dtype= tf.float64, seed= 0))

In [19]:
#날코딩
y= 1/(1+ np.e**(a* x_data + b))

#### 짠걸로 로지스틱 회귀 돌려봄
결과를 0 or 1로 만들었으니 loss가 확 줄어들 수 밖에

In [20]:
loss= -tf.reduce_mean(np.array(y_data) * tf.log(y) + (1- np.array(y_data)) * tf.log(1-y))

In [21]:
learning_rate= 0.5

In [23]:
# 경사하강. loss를 최소로하는
gradient_decent= tf.train.GradientDescentOptimizer(learning_rate).minimize(loss)

In [25]:
#학습
with tf.Session() as ss:
    ss.run(tf.global_variables_initializer())
    
    for i in range(60001):
        ss.run(gradient_decent)
        if 0== i%6000:
            print('epoch%12.6f, loss%10.6f, 기울기a%10.6f, y절편b%10.6f'%
                 (i, ss.run(loss), ss.run(a), ss.run(b)))

epoch    0.000000, loss  1.267648, 기울기a  0.184897, y절편b -0.433382
epoch 6000.000000, loss  0.015193, 기울기a -2.921086, y절편b 20.298153
epoch12000.000000, loss  0.008070, 기울기a -3.563739, y절편b 24.801002
epoch18000.000000, loss  0.005475, 기울기a -3.955722, y절편b 27.546315
epoch24000.000000, loss  0.004137, 기울기a -4.238023, y절편b 29.523131
epoch30000.000000, loss  0.003323, 기울기a -4.458593, y절편b 31.067548
epoch36000.000000, loss  0.002775, 기울기a -4.639561, y절편b 32.334608
epoch42000.000000, loss  0.002382, 기울기a -4.792965, y절편b 33.408639
epoch48000.000000, loss  0.002086, 기울기a -4.926080, y절편b 34.340596
epoch54000.000000, loss  0.001855, 기울기a -5.043638, y절편b 35.163619
epoch60000.000000, loss  0.001671, 기울기a -5.148889, y절편b 35.900470


#### 로지스틱회귀: 여러입력값

In [26]:
import numpy as np

In [27]:
seed= 0
np.random.seed(seed)
tf.set_random_seed(seed)

In [28]:
x_data= np.array([[2,3], [4,3], [6,4], [8,6], [10,7], [12,8], [14,9]])
y_data= np.array([0, 0, 0, 1, 1, 1, 1]).reshape(7,1)

입력값을 플레이스홀더에 저장
- placeholder ??

In [29]:
X= tf.placeholder(tf.float64, shape= [None, 2])
Y= tf.placeholder(tf.float64, shape= [None, 1])

In [30]:
# 임의 초깃값
a= tf.Variable(tf.random_uniform([2,1], dtype= tf.float64)) #[들어오는 2개, 나가는 1개 값]
b= tf.Variable(tf.random_uniform([1], dtype= tf.float64))

기 정의된 시그모이드 함수

In [31]:
y= tf.sigmoid(tf.matmul(X, a) + b) #matmul: 행렬곱

In [32]:
# 오차
loss= -tf.reduce_mean(Y* tf.log(y) + (1-Y)* tf.log(1-Y))

In [34]:
# 학습률. loss줄이려고 얼마씩 이동?변경?할지
learning_rate= 0.1

In [37]:
# 오차 최소. 경사하강법
gradient_descent= tf.train.GradientDescentOptimizer(learning_rate).minimize(loss)

#### 애매한 중간값 없이, threshold 마냥 확연한 구분!

In [38]:
predicted= tf.cast(y> 0.5, dtype= tf.float64)

In [39]:
# reduce_mean: 그냥 평균인데, python 기본함수랑 구분하려고
accuracy= tf.reduce_mean(tf.cast( tf.equal(predicted, Y),
                                dtype= tf.float64))

학습

In [41]:
with tf.Session() as ss:
    ss.run(tf.global_variables_initializer())
    
    for i in range(3001):
        a_, b_, loss_, _ = ss.run([a, b, loss, gradient_descent],
                                 feed_dict= {X: x_data, Y: y_data})
        if 0== (i+1)%300:
            print('step%5d, a1%10.6f, a2%10.6f, b%10.6f, loss%10.6f'%
                 (i+1, a_[0], a_[1], b_, loss_))

step  300, a1  0.924412, a2  0.163625, b  0.798926, loss       nan
step  600, a1  0.928549, a2  0.166690, b  0.799424, loss       nan
step  900, a1  0.932473, a2  0.169598, b  0.799897, loss       nan
step 1200, a1  0.936206, a2  0.172364, b  0.800347, loss       nan
step 1500, a1  0.939765, a2  0.175002, b  0.800776, loss       nan
step 1800, a1  0.943165, a2  0.177522, b  0.801186, loss       nan
step 2100, a1  0.946421, a2  0.179936, b  0.801579, loss       nan
step 2400, a1  0.949544, a2  0.182252, b  0.801956, loss       nan
step 2700, a1  0.952545, a2  0.184477, b  0.802319, loss       nan
step 3000, a1  0.955432, a2  0.186618, b  0.802667, loss       nan
